<?xml version="1.0"?>
<launch>
    <arg name="offline" default="true" />

    <arg name="map" value="$(optenv MAP lego_loam_normals_map)"/>
    <arg name="map_file" default="$(find mcl_standard)/maps/$(arg map).pcd"/>
    <arg name="num_particles" default="3" />
    <arg name="outlier_weight" default="0.0" />
    <arg name="obs_std_dev" default="0.02" />
    
    <group if="$(arg offline)">
        <param name="/use_sim_time" value="true" />
        <node pkg="tf" type="static_transform_publisher" name="base_link_to_rslidar"  args="0 0 1.038 0 0 0 /base_footprint /rslidar 1" />
        <node pkg="mcl_standard" type="odom_broadcaster_node" name="odom_broadcaster_node" output="screen"/>
        <node pkg="rviz" type="rviz" name="rviz" args="-d $(find mcl_standard)/config/mcl_standard.rviz" />
    </group>

    <node pkg="pcl_ros" type="pcd_to_pointcloud" name="pcd_to_pointcloud" args="$(arg map_file)">
        <param name="frame_id" value="map" />
        <remap from="cloud_pcd" to="mapcloud" />
    </node>

    <node pkg="mcl_standard" type="mcl_standard_node"    name="mcl_standard_node"    output="screen">
        <param name="map_file" value="$(arg map_file)" />
        <param name="num_particles" value="$(arg num_particles)" />
        <param name="obs_std_dev" value="$(arg obs_std_dev)" />
        <param name="outlier_weight" value="$(arg outlier_weight)" />
        <param name="init_trans_var" value="0.1" />
        <param name="init_rot_var" value="0" />
        <!--
        <param name="odom_trans_var_per_m" value="1" />
        -->
    </node>

    <node pkg="normal_estimation" type="approx_normal_estimation_node" name="normal_estimation_node" output="screen">
        <remap from="lidar_points" to="rslidar_points" />
        <param name="max_row_nn_dist" value="0.07" />
        <param name="max_nn_minor_axis_var" value="0.00005" />
        <param name="min_nn_major_axes_var" value="0.0009" />
        <param name="max_pt_to_normal_ang" value="1.3" />
    </node>
</launch>
